src/plugins/workspace/public/components/data_source_association/association_data_source_modal.tsx

```diff
@@ -229,28 +241,79 @@ export const AssociationDataSourceModalContent = ({
   }, [selectedConnectionIds, allConnections, handleAssignDataSourceConnections]);
 
   useEffect(() => {
-    setIsLoading(true);
-    getDataSourcesList(savedObjects.client, ['*'])
-      .then((dataSourcesList) => fetchDataSourceConnections(dataSourcesList, http, notifications))
-      .then((connections) => {
-        setAllConnections(connections);
-      })
-      .finally(() => {
+    const fetchDataSources = async () => {
+      const dataSourcesList = await getDataSourcesList(savedObjects.client, ['*']);
+      const {
+        openSearchConnections,
+        dataConnections,
+      } = convertDataSourcesToOpenSearchAndDataConnections(dataSourcesList);
+      return { openSearchConnections, dataConnections };
+    };
+
+    const handleDirectQueryConnections = async (
+      openSearchConnections: DataSourceConnection[],
+      dataConnections: DataSourceConnection[]
+    ) => {
+      if (mode === AssociationDataSourceModalMode.OpenSearchConnections) {
+        setAllConnections([...openSearchConnections, ...dataConnections]);
         setIsLoading(false);
-      });
-  }, [savedObjects.client, http, notifications, mode]);
+        return;
+      }
+
+      // When mode is  Direct Query connections
+      const fetchDqcConnectionsPromises = openSearchConnections.map((ds) =>
+        fetchDirectQueryConnectionsByIDs([ds.id], http, notifications)
+          .then((directQueryConnections) => ({
+            id: ds.id,
+            relatedConnections: directQueryConnections,
+          }))
+          .catch(() => ({
+            id: ds.id,
+            relatedConnections: [],
+          }))
+      );
+      const dqcConnections = await Promise.all(fetchDqcConnectionsPromises);
+
+      const allConnectionsWithDQC = openSearchConnections
+        .filter((connection) => {
+          const filteredData = dqcConnections.find((c) => c.id === connection.id);
+          return filteredData && filteredData.relatedConnections.length > 0;
+        })
+        .map((connection) => {
+          const relatedDQC = dqcConnections.find((c) => c.id === connection.id);
+          return {
+            ...connection,
+            relatedConnections: relatedDQC?.relatedConnections,
+          } as DataSourceConnection;
+        });
+
+      setAllConnections(allConnectionsWithDQC);
+      setIsLoading(false);
+    };
+
+    const fetchDataSourcesAndHandleRelatedConnections = async () => {
+      setIsLoading(true);
+      const { openSearchConnections, dataConnections } = await fetchDataSources();
+      await handleDirectQueryConnections(openSearchConnections, dataConnections);
```

Comment:
```suggestion
      const { allConnections } = await handleDirectQueryConnections(openSearchConnections, dataConnections);
      setAllConnections(allConnections);
      setIsLoading(false);
```
I feel it will be more clear that we make the function stateless, and we can move the function out of the component to increase the readability of the component and the useEffect hook.

Reply:
Thanks for the comments, I will update it.